### PR TITLE
feat(e2e) Detox testing

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -5,21 +5,27 @@ import org.gradle.util.VersionNumber
...
+
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.10"
     }
 }
 
 ext {
     buildToolsVersion = "30.0.3"
     compileSdkVersion = 31
-    minSdkVersion    = 23
+    minSdkVersion    = 26
```
Comment (saghul) [inline android/build.gradle:28]: We cannot do this, that would cut support for many old-but-still-used Android versions.

File: .detoxrc.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            "device": {
+                "type": "iPhone 13 Pro"
+            }
+        }
+    },
+    "apps": {
+        "android.debug": {
+            "type": "android.apk",
+            "binaryPath": "android/app/build/outputs/apk/debug/app-debug.apk",
+            "build": "cd android && ./gradlew assembleDebug assembleAndroidTest -DtestBuildType=debug && cd .."
```
Comment (saghul) [inline .detoxrc.json:24]: We should try to make a release build, we can do that by un-commenting a line in the Gradle file so the app is signed with a test key, which is ok.

File: android/app/build.gradle [hunk 3]
```diff
@@ -31,6 +31,11 @@ android {
         ndk {
             abiFilters 'armeabi-v7a', 'arm64-v8a', 'x86', 'x86_64'
         }
+
+        // This will later be used to control the test apk build type
+        testBuildType System.getProperty('testBuildType', 'debug')
```
Comment (saghul) [inline android/app/build.gradle:36]: Where is this used? I see it in the build command line, is that all? It's also se to the same value.

File: android/app/src/androidTest/java/org/jitsi/meet/DetoxTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+import androidx.test.filters.LargeTest;
+import androidx.test.rule.ActivityTestRule;
+
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class DetoxTest {
```
Comment (saghul) [inline android/app/src/androidTest/java/org/jitsi/meet/DetoxTest.java:16]: Is this needed as some kind of pre-requisite for Detox to run?

File: android/build.gradle [hunk 5]
```diff
@@ -5,21 +5,27 @@ import org.gradle.util.VersionNumber
...
         google()
         jcenter()
     }
     dependencies {
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion"
         classpath 'com.android.tools.build:gradle:4.1.2'
         classpath 'com.google.gms:google-services:4.3.4'
         classpath 'com.google.firebase:firebase-crashlytics-gradle:2.4.1'
+
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.6.10"
```
Comment (saghul) [inline android/build.gradle:21]: This dependency is duplicated. Also, is it necessary?

File: e2e/MuteAudioVideoTest.e2e.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    beforeAll(async () => {
+        await device.launchApp({
+            newInstance: true,
+            permissions: {
+                camera: 'YES',
+                microphone: 'YES'
+            }
+        });
+
+        await device.setURLBlacklist([
```
Comment (saghul) [inline e2e/MuteAudioVideoTest.e2e.js:13]: Why are you setting these?

File: e2e/ChangeDisplayNameTest.e2e.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+// @flow
+
+import { screen } from '../react/features/conference/components/native/routes';
+
+describe('Jitsi Meet App', () => {
```
Comment (saghul) [inline e2e/ChangeDisplayNameTest.e2e.js:5]: Let's give each test a more descriptive name?

**Review by saghul — COMMENTED** [review]
Looks like a great start! What is your overall feeling when working with Detox?

File: android/app/src/androidTest/java/org/jitsi/meet/DetoxTest.java (hunk 4 above)
Comment (Calinteodor) [reply android/app/src/androidTest/java/org/jitsi/meet/DetoxTest.java:16]: Yes, it's part of the Android setup.

File: android/build.gradle (hunk 5 above)
Comment (Calinteodor) [reply android/build.gradle:21]: Sorry, will remove duplication.

File: android/build.gradle (hunk 1 above)
Comment (Calinteodor) [reply android/build.gradle:28]: image

This was the reason behind this change. I cannot build the app unless version is 26.

File: android/app/build.gradle (hunk 3 above)
Comment (Calinteodor) [reply android/app/build.gradle:36]: Yes. That's the only place where it's used.

File: e2e/MuteAudioVideoTest.e2e.js (hunk 6 above)
Comment (Calinteodor) [reply e2e/MuteAudioVideoTest.e2e.js]: Removed. No longer needed.

File: e2e/ChangeDisplayNameTest.e2e.js (hunk 7 above)
Comment (Calinteodor) [reply e2e/ChangeDisplayNameTest.e2e.js]: Moved everything to WelcomePageTest.e2e.js.

File: android/build.gradle (hunk 1 above)
Comment (Calinteodor) [reply android/build.gradle]: This is now reversed to 23.

**Comment by saghul** [comment]
I have bad news. The reason why Detox fails on Android is because our app deviates from the standard RN app and doesn't quite meet Detox's expectations, as it behaves like a "brownfield" app.

Here is the output:

~~~
02-09 13:03:16.038 23240 23266 I Detox   : Detox server connection details: url=ws://localhost:52164, sessionId=d5159cdb-2ed9-e870-1a29-fa0704c3e747
02-09 13:03:16.067 23240 23299 I DetoxDispatcher: Handling action '_init' (ID #0)...
02-09 13:03:16.067 23240 23299 I Detox   : Connecting to server...
02-09 13:03:16.067 23240 23299 I DetoxWSClient: At connectToServer
02-09 13:03:16.093 23240 23299 E AndroidRuntime: FATAL EXCEPTION: detox.primary
02-09 13:03:16.093 23240 23299 E AndroidRuntime: Process: org.jitsi.meet, PID: 23240
02-09 13:03:16.093 23240 23299 E AndroidRuntime: java.lang.ClassCastException: android.app.Application cannot be cast to com.facebook.react.ReactApplication
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.reactnative.ReactNativeExtension.waitForRNBootstrap(ReactNativeExtension.kt:58)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.DetoxMain.initReactNativeIfNeeded(DetoxMain.kt:102)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.DetoxMain.doInit(DetoxMain.kt:34)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.DetoxMain.access$doInit(DetoxMain.kt:15)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.DetoxMain$initActionHandlers$1$1.handle(DetoxMain.kt:52)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.ActionsExecutor.executeAction$lambda-2$lambda-1(DetoxActionsDispatcher.kt:64)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.ActionsExecutor.lambda$nk3cF8ntnfFqbJOcVqd04vjb5nY(Unknown Source:0)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.-$$Lambda$ActionsExecutor$nk3cF8ntnfFqbJOcVqd04vjb5nY.run(Unknown Source:8)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:223)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.ActionsExecutor._init_$lambda-0(DetoxActionsDispatcher.kt:50)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.ActionsExecutor.lambda$Hi5vyewqQiQqk57AkJy_cHqNGKY(Unknown Source:0)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at com.wix.detox.adapters.server.-$$Lambda$ActionsExecutor$Hi5vyewqQiQqk57AkJy_cHqNGKY.run(Unknown Source:4)
02-09 13:03:16.093 23240 23299 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:923)
02-09 13:03:16.093 23240 23299 E DetoxCrashHandler: Crash detected!!! thread=detox.primary (1126)
~~~

Detox themselves want to eventually fix this, but we have no indication as of when: https://github.com/wix/Detox/issues/2543

**Comment by Calinteodor** [comment]
We need to find a way to not use JitsiMeetCompanion when running tests on iOS. 
Build fails otherwise.

**Comment by saghul** [comment]
We need to build with bitcode, that should do.

**Comment by Calinteodor** [comment]
Build Succeeded on iOS! Tests pass! This is good to go @saghul!

**Comment by saghul** [comment]
Exciting!

File: tests/mobile/WelcomePageTest.e2e.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/* eslint-disable no-undef */
```
Comment (saghul) [inline tests/mobile/WelcomePageTest.e2e.js:1]: Try not to do this and use `/* global foo */ instead.

File: tests/mobile/WelcomePageTest.e2e.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        await device.launchApp({
+            newInstance: true,
+            permissions: {
+                camera: 'YES',
+                microphone: 'YES'
+            }
+        });
+
+        // eslint-disable-next-line no-unused-expressions
+        true;
```
Comment (saghul) [inline tests/mobile/WelcomePageTest.e2e.js:20]: I think you can just do `return true;` to tame the linter.

File: tests/mobile/WelcomePageTest.e2e.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    });
+
+    beforeEach(async () => {
+        await device.reloadReactNative();
```
Comment (saghul) [inline tests/mobile/WelcomePageTest.e2e.js:24]: We don't want this because we run in release mode.

File: tests/mobile/WelcomePageTest.e2e.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    });
+
+    it('is welcome screen enabled', async () => {
+        await expect(element(by.id('welcome-screen'))).toBeVisible();
+    });
+
+    it('did the display name update', async () => {
+        const displayName = 'Benedetta';
+
+        await element(by.id('toggle-drawer-button')).tap();
```
Comment (saghul) [inline tests/mobile/WelcomePageTest.e2e.js:34]: Pl define some constants.

File: tests/mobile/WelcomePageTest.e2e.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        await device.disableSynchronization();
+        await element(by.id('room-name-input')).tap();
+        await element(by.id('room-name-input')).typeText(roomName);
+        await element(by.id('join-room-button')).tap();
+
+        // Conference joined
+        await sleep(2000);
+        await expect(element(by.label('Mute / Unmute'))).toBeVisible();
+        await expect(element(by.label('Start / Stop camera'))).toBeVisible();
+        await element(by.label('Mute / Unmute')).tap();
```
Comment (saghul) [inline tests/mobile/WelcomePageTest.e2e.js:61]: Try to use an ID

File: tests/mobile/config.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+    "maxWorkers": 1,
+    "testEnvironment": "./environment",
+    "testRunner": "jest-circus/runner",
+    "testTimeout": 120000,
+    "testRegex": "\\.e2e\\.js$",
```
Comment (saghul) [inline tests/mobile/config.json:6]: LEt's call our tests Foo.test.js

**Review by saghul — CHANGES_REQUESTED** [review]

**Pull request closed without merging**